File: test/leetcode/m30/SubstringWithConcatenationOfAllWordsTest.java
/*
 * Copyright (c) 2024-2025 THGEEK.COM. All rights reserved.
 */

package leetcode.m30;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import support.ArrayUtils;

import java.util.List;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Test for {@link SubstringWithConcatenationOfAllWords}
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/08/07 11:18
 */
class SubstringWithConcatenationOfAllWordsTest {
    private SubstringWithConcatenationOfAllWords test;

    @BeforeEach
    void setUp() {
        test = new SubstringWithConcatenationOfAllWords();
    }

    @Test
    void test1() {
        String s = "barfoothefoobarman";
        String[] words = {"foo", "bar"};
        List<Integer> expected = ArrayUtils.convert(new Integer[]{0, 9});
        assertEquals(expected, test.findSubstring(s, words));
    }

    @Test
    void test2() {
        String s = "wordgoodgoodgoodbestword";
        String[] words = {"word", "good", "best", "word"};
        List<Integer> expected = ArrayUtils.convert(new Integer[]{});
        assertEquals(expected, test.findSubstring(s, words));
    }

    @Test
    void test3() {
        String s = "barfoofoobarthefoobarman";
        String[] words = {"bar", "foo", "the"};
        List<Integer> expected = ArrayUtils.convert(new Integer[]{6, 9, 12});
        assertEquals(expected, test.findSubstring(s, words));
    }

    @Test
    void test4() {
        String s = "wordgoodgoodgoodbestword";
        String[] words = {"word", "good", "best", "good"};
        List<Integer> expected = ArrayUtils.convert(new Integer[]{8});
        assertEquals(expected, test.findSubstring(s, words));
    }

    @Test
    void test5() {
        String s = "aaaaaaaaaaaaaa";
        String[] words = {"aa", "aa"};
        List<Integer> expected = ArrayUtils.convert(new Integer[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10});
        List<Integer> actual = test.findSubstring(s, words).stream().sorted().collect(Collectors.toList());
        assertEquals(expected, actual);
    }
}

File: src/leetcode/m30/SubstringWithConcatenationOfAllWords_test.go
package m30

import (
	"reflect"
	"testing"
)

func Test_findSubstring(t *testing.T) {
	type args struct {
		s     string
		words []string
	}
	tests := []struct {
		name string
		args args
		want []int
	}{
		{"Test case 1", args{"barfoothefoobarman", []string{"foo", "bar"}}, []int{0, 9}},
		{"Test case 2", args{"wordgoodgoodgoodbestword", []string{"word", "good", "best", "word"}}, []int{}},
		{"Test case 3", args{"barfoofoobarthefoobarman", []string{"bar", "foo", "the"}}, []int{6, 9, 12}},
		{"Test case 4", args{"lingmindraboofooowingdingbarrwingmonkeypoundcake", []string{"fooo", "barr", "wing", "ding", "wing"}}, []int{13}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := findSubstring(tt.args.s, tt.args.words); !reflect.DeepEqual(got, tt.want) {
				t.Errorf("findSubstring() = %v, want %v", got, tt.want)
			}
		})
	}
}
